src/ripple/rpc/NFTokenID.h

```diff
@@ -0,0 +1,65 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_RPC_NFTOKENID_H_INCLUDED
+#define RIPPLE_RPC_NFTOKENID_H_INCLUDED
+
+#include <ripple/protocol/Protocol.h>
+
+#include <functional>
+#include <memory>
+
+namespace Json {
+class Value;
+}
+
+namespace ripple {
+
+class TxMeta;
+class STTx;
+
+namespace RPC {
+
+/**
+   Add a `nftoken_ids` field to the `meta` output parameter.
+   The field is only added to successful NFTokenMint, NFTokenAcceptOffer,
+   and NFTokenCancelOffer transactions.
+
+   @{
+ */
+bool
+canHaveNFTokenID(std::shared_ptr<STTx const> const&, TxMeta const&);
+
+void
+getNFTokenIDFromPage(TxMeta const&, std::vector<uint256>&);
+
+void
+getNFTokenIDFromDeletedOffer(TxMeta const&, std::vector<uint256>&);
```

Comment:
what benefit would do we get from limiting the visibility to just the translation unit? is it just for good practice?

Reply:
There actually is a benefit, but it may be a little hard to understand.

C++ (and C) compilers and linkers work by identifying things (variables, chunks of code) by name.  Then it resolves uses of those things by name.

When a function is labeled `static` in a cpp file at namespace scope, the compiler uses the (mangled) name of that function for resolution within the transaction unit.  But when the compiler builds the object file it does _not_ publish any namespace scoped `static` names so they can be used by the linker.

This has a couple of benefits.

1. If the function is `static` we can be sure it is only being used within the translation unit.  If I want to find all callers of the function, then I can be confident that simply by looking for uses within the translation unit I can find them all.  That can be useful for refactoring.

2. Information hiding and privacy are important ideas in C++.  You may think that, for example, putting a declaration in a `private:` section of a `class` makes it private.  But, in fact, by placing the declaration in the header you have published to the world (anyone who can include the header) the existence of the function.  Any `friend`s of that class can call the private function.  Information that has been made available can become depended upon by various means.  So a `static` function at namespace scope is, in many ways, more private than a class method that has been declared `private`.

Let's suppose we just took the function declaration out of the header, and left the implementation in the cpp file, but did not mark the function `static`.  That seems like it should be just as effective.  But it's not.

It is possible for some well intentioned engineer taking a short cut to forward declare the function _anywhere_, not just in the appropriate header file.  Once that forward declaration is visible, the (non-`static`) function is magically callable from anywhere in the program -- not just in the transaction unit.

However, if we take the time to declare the function `static` the forward declaration does no good (outside of the translation unit containing the function implementation).  The `static` removes from the linker the ability to resolve the name.  Someone who attempts to call the function gets a linker error about something being unresolved.

Does that explanation make sense?  Yes, it's good practice.  But it's also good practice for a reason.  Information hiding is valuable.  This is some of the best information hiding that the language has.

Reply:
Thanks so much for the detailed explanation!!! Makes sense to hide the helper function